tests/xrt/utils.cmake

```diff
@@ -0,0 +1,144 @@
+  endif()
+endif()
+
+if(NOT XRT)
+  find_package(XRT)
+endif()
+
+set(xrt_core_LIBRARY XRT::xrt_core)
+set(xrt_coreutil_LIBRARY XRT::xrt_coreutil)
+set(xrt_xilinxopencl_LIBRARY XRT::xilinxopencl)
+set(xrt_core_static_LIBRARY XRT::xrt_core_static)
+set(xrt_coreutil_static_LIBRARY XRT::xrt_coreutil_static)
+set(xrt_xilinxopencl_static_LIBRARY XRT::xilinxopencl_static)
+
+if (DEFINED ENV{XCL_EMULATION_MODE})
+    set(MODE $ENV{XCL_EMULATION_MODE})
+    string(REPLACE "_" "" XCL_EMU_SUFFIX ${MODE})
+    set(xrt_core_LIBRARY XRT::xrt_${XCL_EMU_SUFFIX})
+endif()
+
+include_directories(${XRT_INCLUDE_DIRS})
+
+if (NOT WIN32)
+  if(NOT DEFINED uuid_LIBRARY)
+    find_library(uuid_LIBRARY
+    NAMES uuid)
+    message("uuid_LIBRARY=${uuid_LIBRARY}")
+  endif()
+else()
+    set(BOOST_ROOT C:/Xilinx/XRT/ext)
+    find_package(Boost)
+    include_directories(${Boost_INCLUDE_DIRS})
+endif(NOT WIN32)
+
+if (NOT DEFINED ${INSTALL_DIR})
+  set(INSTALL_DIR "${CMAKE_CURRENT_SOURCE_DIR}/build/${CMAKE_SYSTEM_NAME}/${CMAKE_BUILD_TYPE}/${MODE}")
+endif()
+
+# function for creation emconfig.json file for sw_emu, hw_emu
+function(xrt_create_emconfig PLATFORM)
+  add_custom_command(
+    OUTPUT emconfig.json
+    COMMAND emconfigutil
+    --platform ${PLATFORM}
+    --nd 1
+  )
+  add_custom_target(EMCONFIG_${TESTNAME}
+  ALL
+  DEPENDS emconfig.json
+  )   
+endfunction()
+
+# macro to generate XO file
+# xrt_create_xo macro has two arguments, SRC, OUT
+# SRC                     Path to kernel code (.cl | .cpp)
+# OUT                     Name of the .xo to be generated
+macro(xrt_create_xo SRC OUT)
+  set(OUT_FILE "${OUT}_${MODE}")
+  set(OUT_TARGET "${OUT_FILE}_${TESTNAME}")
+  list(APPEND XOS "${CMAKE_CURRENT_BINARY_DIR}/${OUT_FILE}.xo")
+  list(APPEND XO_TARGETS "${OUT_TARGET}")
+  add_custom_command(
+    OUTPUT ${OUT_FILE}.xo
+    COMMAND ${VITIS_COMPILER}
+    -c ${SRC}
+    ${EXTRA_COMPILE_OPTIONS}
+    -o ${CMAKE_CURRENT_BINARY_DIR}/${OUT_FILE}.xo
+    --platform ${PLATFORM}
+    -t ${MODE}
+  )
+  add_custom_target(${OUT_TARGET}
+    ALL
+    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${OUT_FILE}.xo
+  )
+endmacro()
+
+# macro to generate xclbin file
+# xrt_create_xclbin() macro has four arguments, CMD, EXTRA_LINK_OPTIONS, OUT, XO_TARGET
+# LINK_OPTIONS    Any extra link options while generating xclbin to be provided
+# OUT             Name of the xclbin to be generated 
+# KERNEL_SRC      List of Kernel source files required for generating xclbin
+macro(xrt_create_xclbin)
+  
+  # cmake_parse_arguments general structure is :
+  # cmake_parse_arguments(SUFFIX "${options}" "${oneValueArg}" "${multipleValueArg}" ..)
+  # Here, OUT and LINK_OPTIONS can have only one arg
+  # KERNEL_SRC can have multiple arguments. This is useful in testcases where there are multiple src files.
+  set(one_value_arg OUT LINK_OPTIONS)
+  cmake_parse_arguments(XCLBIN "" "${one_value_arg}" KERNEL_SRC "${ARGN}")
```

Comment:
This looks pretty complex. Instead of this i would suggest you to go back and have create_xo for each kernel and create_xclbin for xclbin creation.

Reply:
cc: @stsoe , Please suggest. It is better to have two functions instead of this complex parsing in one function to figure out which one for compilation and which one is for linking etc.
create_xo(kernel_src, extra_comp_args)
create_xclbin(kernel_xos, extra_link_args)